File: 0224-basic-calculator/0224-basic-calculator.py
class Parser:
    def __init__(self):
        self.index = 0
        self.value = 0
        self.s = ""

    def parse(self, s):
        self.s = s
        self.index = 0
        return self.expr()

    def _skip_spaces(self):
        while self.index < len(self.s) and self.s[self.index] == " ":
            self.index += 1

    def _apply(self, op, a, b):
        if op == "+":
            return a + b
        if op == "-":
            return a - b
        if op == "*":
            return a * b
        if op == "/":
            return a // b
        return 0

    def number(self):
        self._skip_spaces()
        self.value = 0
        while self.index < len(self.s) and self.s[self.index].isdigit():
            self.value = self.value * 10 + int(self.s[self.index])
            self.index += 1
        return self.value

    def factor(self):
        self._skip_spaces()
        if self.index < len(self.s) and self.s[self.index] == "(":
            self.index += 1
            val = self.expr()
            self._skip_spaces()
            if self.index < len(self.s) and self.s[self.index] == ")":
                self.index += 1 
            return val
        return self.number()

    def term(self):
        result = self.factor()
        self._skip_spaces()
        while self.index < len(self.s) and self.s[self.index] in "*/":
            op = self.s[self.index]
            self.index += 1
            result = self._apply(op, result, self.factor())
            self._skip_spaces()
        return result

    def expr(self):
        result = self.term()
        self._skip_spaces()
        while self.index < len(self.s) and self.s[self.index] in "+-":
            op = self.s[self.index]
            self.index += 1
            result = self._apply(op, result, self.term())
            self._skip_spaces()
        return result

class Solution(object):
    def calculate(self, s):
        solver = Parser()
        return solver.parse(s)

File: 0224-basic-calculator/0224-basic-calculator.cpp
class Parser {
    using ll = long long;

    string s;
    size_t i = 0;

    void skipSpaces() {
        while (i < s.size() && s[i] == ' ') {
            i++;
        }
    }

    bool hasMore() {
        skipSpaces();
        return i < s.size();
    }

    char peek() {
        skipSpaces();
        return s[i];
    }

    char consume() {
        skipSpaces();
        return s[i++];
    }

    ll number() {
        skipSpaces();

        ll n = 0;

        while (i < s.size() && isdigit(s[i])) {
            n = n * 10 + (s[i] - '0');
            i++;
        }

        return n;
    }

    ll factor() {
        skipSpaces();

        if (peek() == '+') {
            consume();
            return factor();
        }

        if (peek() == '-') {
            consume();
            return -factor();
        }

        if (peek() == '(') {
            consume();          // '('
            ll val = expr();
            consume();          // ')'
            return val;
        }

        return number();
    }

    ll term() {
        ll result = factor();

        while (hasMore() && (peek() == '*' || peek() == '/')) {
            char op = consume();
            ll right = factor();

            if (op == '*') {
                result *= right;
            } else {
                result /= right;
            }
        }

        return result;
    }

    ll expr() {
        ll result = term();

        while (hasMore() && (peek() == '+' || peek() == '-')) {
            char op = consume();
            ll right = term();

            if (op == '+') {
                result += right;
            } else {
                result -= right;
            }
        }

        return result;
    }

public:
    Parser(const string& expression) : s(expression) {}

    ll eval() {
        return expr();
    }
};

class Solution {
public:
    int calculate(string s) {
        Parser p(s);
        return (int)p.eval();
    }
};
